### PR TITLE
[SPARK-23099][SS] Migrate foreach sink to DataSourceV2

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Migrate the foreach sink to the DataSourceV2 API.

## How was this patch tested?

existing unit tests, and new test to verify edge case

### CODE REVIEW

**Review by tdas — CHANGES_REQUESTED** [review]
This is fine, just a few comments. But one major concern I have is the way we are indirectly exposing the batchId to the DataWriter (using job properties). If we are not exposing it directly, then essentially we are saying is that other data source build by general developers are not supposed to use batch ids in the executors for any purpose. In a way that is very odd: there is some transactional logic that you can write with ForeachWriter (which exposes the id), but you cannot write the same logic with DataSource V2 APIs. 

I know that the alternative is also not great; we will have to again split the StreamWriter to separate APIs for MicroBatch and Continuous.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 1]
```diff
@@ -17,52 +17,119 @@
...
-        writer.close(null)
-      } else {
-        writer.close(null)
+      schema.toAttributes,
+      SparkSession.getActiveSession.get.sessionState.analyzer)
+    ForeachInternalWriter(writer, encoder)
+  }
+}
+
+case class ForeachInternalWriter[T: Encoder](
```
[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: actually.. probably should not inline this. its outer closure may not be serializable in that case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 2]
```diff
@@ -17,52 +17,119 @@
...
+    extends StreamWriter with SupportsWriteInternalRow {
+  override def commit(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {}
+  override def abort(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {}
+
+  override def createInternalRowWriterFactory(): DataWriterFactory[InternalRow] = {
+    ForeachWriterFactory(writer, encoder)
+  }
+}
+
+case class ForeachWriterFactory[T: Encoder](writer: ForeachWriter[T], encoder: ExpressionEncoder[T])
```
[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: actually.. probably should not inline this. its outer closure may not be serializable in that case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 3]
```diff
@@ -17,52 +17,119 @@
...
+    val CLOSED, OPENED, OPENED_SKIP_PROCESSING = Value
+  }
+  import WriterState._
+
+  private var state = CLOSED
+
+  private def openAndSetState(epochId: Long) = {
+    // Create a new writer by roundtripping through the serialization for compatibility.
+    // In the old API, a writer instantiation would never get reused.
+    val byteStream = new ByteArrayOutputStream()
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: You're right; this suggestion is what we really want.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 4]
```diff
@@ -255,6 +255,32 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
       query.stop()
     }
   }
+
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: Good instinct, it didn't quite work. Added the test.

**Comment by jose-torres** [comment]
It's my intent to say that other data sources built by general developers aren't supposed to use batch ids in the executors for any purpose. In addition to the issue you mentioned, I don't think there's a compelling reason to do so in the DataSourceV2 model, and I worry it's easy to write implementations that seem correct but aren't that way.

Since this interface is still evolving, I think it makes sense to revisit the question if we notice a scenario where it's infeasible to rewrite a piece of transactional logic to not use the batch ID in the executor.

**Comment by jose-torres** [comment]
Filed SPARK-23416 for the unrelated failure in build 87241.

**Comment by jose-torres** [comment]
This is obsolete - we're changing the lifecycle of DataWriter.

**Pull request closed without merging**